### PR TITLE
[SPARK-30659][ML][PYSPARK] LogisticRegression blockify input vectors

### PR DESCRIPTION
**Pull request by zhengruifeng**
### What changes were proposed in this pull request?
1, use blocks instead of vectors
2, use Level-2 BLAS for binary, use Level-3 BLAS for multinomial 

### Why are the changes needed?
1, less RAM to persist training data; (save ~40%)
2, faster than existing impl; (40% ~ 92%)

### Does this PR introduce any user-facing change?
add a new expert param `blockSize`

### How was this patch tested?
updated testsuites

### CODE REVIEW

**Comment by zhengruifeng** [comment]
env: bin/spark-shell --driver-memory=32G

testCode:
```scala

import org.apache.spark.ml.classification._
import org.apache.spark.storage.StorageLevel


var df = spark.read.format("libsvm").load("/data1/Datasets/a9a/a9a").withColumn("label", (col("label")+1)/2)
df.persist(StorageLevel.MEMORY_AND_DISK)
df.count

(0 until 8).foreach{ _ => df = df.union(df) }
df.count

new LogisticRegression().setMaxIter(10).fit(df)

val lr1 = new LogisticRegression().setMaxIter(100).setFamily("binomial")
val start = System.currentTimeMillis; val model1 = lr1.fit(df); val end = System.currentTimeMillis; end - start


val lr2 = new LogisticRegression().setMaxIter(100).setFitIntercept(false).setFamily("binomial")
val start = System.currentTimeMillis; val model2 = lr2.fit(df); val end = System.currentTimeMillis; end - start


val lr3 = new LogisticRegression().setMaxIter(100).setFamily("multinomial")
val start = System.currentTimeMillis; val model3 = lr3.fit(df); val end = System.currentTimeMillis; end - start


val lr4 = new LogisticRegression().setMaxIter(100).setFitIntercept(false).setFamily("multinomial")
val start = System.currentTimeMillis; val model4 = lr4.fit(df); val end = System.currentTimeMillis; end - start
```

result:

this PR:
RAM: 1418.9M
DURATION: 136217, 161194, 171625, 177116

Master:
RAM: 2.3G
DURATION: 217035, 218267, 239111, 250163

File: mllib/src/main/scala/org/apache/spark/ml/optim/aggregator/LogisticAggregator.scala [hunk 1]
```diff
@@ -339,6 +447,116 @@ private[ml] class LogisticAggregator(
...
+    // block.matrix:                  S X F, unknown type
+    // mat (multipliers):             S X C, dense
+    // multinomialLinearGradSumMat:   F X C, dense
+    // gradSumMat(gradientSumArray):  C X FPI (numFeaturesPlusIntercept), dense
+    block.matrix match {
+      case dm: DenseMatrix if !fitIntercept =>
+        // If fitIntercept==false, gradientSumArray += mat.T X matrix
+        // GEMM requires block.matrix is dense
+        val gradSumMat = new DenseMatrix(numClasses, numFeatures, localGradientSumArray)
+        BLAS.gemm(1.0, mat.transpose, dm, 1.0, gradSumMat)
```
Comment (zhengruifeng) [inline mllib/src/main/scala/org/apache/spark/ml/optim/aggregator/LogisticAggregator.scala]: Since `gradientSumArray` is for Matrix of shape CXFPI, and `BLAS.gemm` requires the output matrix is not transposed. So only if F(numFeature) == FPI(numFeaturesPlusIntercept) and input block is dense, can I use `BLAS.gemm` to directly update `gradientSumArray`.
Otherwise, I need to output the result to a temp matrix `multinomialLinearGradSumMat`, and then add elements to `gradientSumArray`

**Comment by zhengruifeng** [comment]
ping @srowen

File: mllib/src/main/scala/org/apache/spark/ml/optim/aggregator/LogisticAggregator.scala [hunk 2]
```diff
@@ -209,6 +208,70 @@ private[ml] class LogisticAggregator(
       s"got type ${bcCoefficients.value.getClass}.)")
   }
 
+  // Helper vectors and matrices for binary:
+  @transient private lazy val binaryLinear = {
```
Comment (srowen) [inline mllib/src/main/scala/org/apache/spark/ml/optim/aggregator/LogisticAggregator.scala]: So, are these lazy just to deal with recreating them after deserialization? they don't seem big, so can they just be non-transient, non-lazy? unless it's a material problem, might be simpler and faster.
Or how much do you need to hold on to scratch vectors like auxiliaryVec vs just locals?

File: mllib/src/test/scala/org/apache/spark/ml/optim/aggregator/LogisticAggregatorSuite.scala [hunk 3]
```diff
@@ -32,21 +32,21 @@ class LogisticAggregatorSuite extends SparkFunSuite with MLlibTestSparkContext {
 
   override def beforeAll(): Unit = {
     super.beforeAll()
-    instances = Array(
+    instances = standardize(Array(
```
Comment (srowen) [inline mllib/src/test/scala/org/apache/spark/ml/optim/aggregator/LogisticAggregatorSuite.scala]: Just curious, why do these have to be standardized now?
One thing I want to be careful about is not introducing bugs of course, and so having it pass existing tests as-is is a good indicator that it doesn't. This may be fine but are we changing inputs to match old outputs - did behavior change?

**Comment by srowen** [comment]
Also, does this cause any appreciable slowdown at smaller scale? it's not a big deal if something that's fast is a little slower, to make things that are slow much faster, but just want to get a sense of what you know about the scale implications.

File: mllib/src/main/scala/org/apache/spark/ml/optim/aggregator/LogisticAggregator.scala (hunk 2 above)
Comment (zhengruifeng) [reply mllib/src/main/scala/org/apache/spark/ml/optim/aggregator/LogisticAggregator.scala]: `binaryLinear`, `binaryIntercept`, `multinomialLinear`, `multinomialIntercept` are the linear and bias part of coefficients, repectively.

`binaryLinearGradSumVec` (numFeatures) and `multinomialLinearGradSumMat` (numClassXnumFeatures) are used to store result of `gemv`/`gemm` if `fitIntercept==True`, since `gradientSumArray` contains gradient sums of intercepts and can not be used directly in `gemv`/`gemm`.

`auxiliaryVec` (blockSize) and `multinomialAuxiliaryMat` (blockSizeXnumClasses) are used to store the intermediate multiplication(margins) and multipliers.

they can be used among blocks, and if they are used multi-times in one call we can assign them to local variables.
However I am OK to make them local variables, since I guess they are not the bottleneck.

File: mllib/src/test/scala/org/apache/spark/ml/optim/aggregator/LogisticAggregatorSuite.scala (hunk 3 above)
Comment (zhengruifeng) [reply mllib/src/test/scala/org/apache/spark/ml/optim/aggregator/LogisticAggregatorSuite.scala]: Two reasons:
1, we do need to standardize the input at each iteration;
2, if we keep this, then it is hard to use matrix operations in aggregator, since we  need: 1, element division, 2, broadcasting (block.matrix / stdvec), 3, special logic that if std==0 then output=0; the first 2 point may need to use `breeze` instead of BLAS, the last point may even make `breeze` not suitable.

**Comment by zhengruifeng** [comment]
@srowen 
The orignial dataset `a9a` is not big, its numFeatures=123, numInstances=32,561, after upsampling its numInstances=32,561X256=8,335,616.

I had made other performance tests, it seems that the performance is related to `numFeatures` and `blockSize`, and I guess the performance is highly related to: given a array of vectors, to what degree can Level2/3-BLAS be faster than existing java impl or Level-1.

Thanks for reviewing!

File: mllib/src/test/scala/org/apache/spark/ml/optim/aggregator/LogisticAggregatorSuite.scala (hunk 3 above)
Comment (srowen) [reply mllib/src/test/scala/org/apache/spark/ml/optim/aggregator/LogisticAggregatorSuite.scala]: This may be a dumb question, but you're not saying that LR input itself has to be standardized first, right? this is just required to meet preconditions of this aggregator? just trying to get a handle on why it wasn't required before. LR now standardizes internally?

**Comment by srowen** [comment]
Yeah that's the question... Level 1 BLAS often isn't a win. L2/L3 yes. It's probably a win, but just dont' want to take a perf hit on most use cases to help large ones. Even that's arguable.

File: mllib/src/test/scala/org/apache/spark/ml/optim/aggregator/LogisticAggregatorSuite.scala (hunk 3 above)
Comment (zhengruifeng) [reply mllib/src/test/scala/org/apache/spark/ml/optim/aggregator/LogisticAggregatorSuite.scala]: Yes, both LinearSVC and LR always standardize the input vector in gradient computation in aggregator.
In existing single-instance `add(instance: Instance)` method, feature value `value` is always used along with std as ` value / localFeaturesStd(index)`

**Comment by zhengruifeng** [comment]
Ok, I will test on small datasets.

**Comment by zhengruifeng** [comment]
@srowen I found that on small datasets, the speed up is even more significant.

data: a9a, numFeatures=123, numInstances=32,561

testCode:
```scala

import org.apache.spark.ml.classification._
import org.apache.spark.storage.StorageLevel


val df = spark.read.format("libsvm").load("/data1/Datasets/a9a/a9a").withColumn("label", (col("label")+1)/2)
df.persist(StorageLevel.MEMORY_AND_DISK)
df.count

val lr4 = new LogisticRegression().setMaxIter(100).setFitIntercept(false).setFamily("multinomial")
val start = System.currentTimeMillis; val model4 = lr4.fit(df); val end = System.currentTimeMillis; end - start

Seq(64, 256, 1024, 4096, 8192).map { b => val start = System.currentTimeMillis; val model1 = new LogisticRegression().setBlockSize(b).fit(df); val end = System.currentTimeMillis; end - start } // this PR

 Seq(64, 256, 1024, 4096, 8192).map { b => val start = System.currentTimeMillis; val model1 = new LogisticRegression().fit(df); val end = System.currentTimeMillis; end - start } // Master

```

result: about **44%~48%** faster. I think that is beacuse on big dataset, the communiation overhead has a bigger impact on the whole procedure; while on small datasets like a9a, high-level BLAS dominates the performance.
This PR: `List(1630, 1623, 1539, 1559, 1666)`
Master: `List(2985, 3037, 2957, 2994, 2959)`

But the way, I set default value to 1024 base on above result. However, the best blocksize will depend on many factors like numFetaures, sparsity, numInstances, etc.

**Comment by zhengruifeng** [comment]
Param `BlockSize` was just added in https://github.com/apache/spark/pull/27360, and only used in LinearSVC and LR, so I can safely chang its default value.

File: mllib/src/main/scala/org/apache/spark/ml/classification/LinearSVC.scala [hunk 4]
```diff
@@ -157,7 +157,7 @@ class LinearSVC @Since("2.2.0") (
 
   /**
    * Set block size for stacking input data in matrices.
-   * Default is 4096.
+   * Default is 1024.
```
Comment (srowen) [inline mllib/src/main/scala/org/apache/spark/ml/classification/LinearSVC.scala]: So this will become the new default for LinearSVC too? OK if so.

File: mllib/src/main/scala/org/apache/spark/ml/optim/aggregator/LogisticAggregator.scala (hunk 2 above)
Comment (srowen) [reply mllib/src/main/scala/org/apache/spark/ml/optim/aggregator/LogisticAggregator.scala]: OK up to your judgment. It'd be simpler to not even make them members, if it's not much difference to performance

File: mllib/src/main/scala/org/apache/spark/ml/optim/aggregator/LogisticAggregator.scala [hunk 5]
```diff
@@ -253,14 +271,69 @@ private[ml] class LogisticAggregator(
...
+    val size = block.size
+    val localGradientSumArray = gradientSumArray
+
+    // vec here represents margins or negative dotProducts
+    val vec = new DenseVector(Array.ofDim[Double](size))
+
+    if (fitIntercept) {
+      val intercept = coefficientsArray.last
+      var i = 0
+      while (i < size) {
```
Comment (srowen) [inline mllib/src/main/scala/org/apache/spark/ml/optim/aggregator/LogisticAggregator.scala]: Would it be faster to fill an array with this value and then make a DenseVector? maybe I'm missing why not

**Review by srowen — COMMENTED** [review]
Looking pretty OK to me

**Comment by srowen** [comment]
I'll merge soon to unblock https://github.com/apache/spark/pull/27389 , but if you have any final thoughts on the above soon, that would be good to check.

File: mllib/src/main/scala/org/apache/spark/ml/classification/LinearSVC.scala (hunk 4 above)
Comment (zhengruifeng) [reply mllib/src/main/scala/org/apache/spark/ml/classification/LinearSVC.scala]: Yes, the new impl of LinearSVC with blocks is quite similar to LR.

File: mllib/src/main/scala/org/apache/spark/ml/optim/aggregator/LogisticAggregator.scala (hunk 5 above)
Comment (zhengruifeng) [reply mllib/src/main/scala/org/apache/spark/ml/optim/aggregator/LogisticAggregator.scala]: OK, I will update it.

**Comment by zhengruifeng** [comment]
since `LeastSquaresAggregator` and `HuberAggregator` in LiR also mark the linear part `coefficients`/`effectiveCoefficientsVector` transient and lazy, so I follow this at the begining.
LiR may also benefit from using blocks instead of vectors, I am working on it.

**Comment by srowen** [comment]
Merged to master

**Comment by WeichenXu123** [comment]
@zhengruifeng Could you provide detail benchmark results separately for:
* dense features
* sparse features
Thanks!

**Comment by mengxr** [comment]
+1 on @WeichenXu123 's suggestion and I would suggest temporarily reverting this change before we have a good solution.

@zhengruifeng @srowen This new approach will introduce significant performance regression on sparse datasets with large number of features, e.g., https://www.csie.ntu.edu.tw/~cjlin/libsvmtools/datasets/binary.html#webspam (16,609,143 features). With block size 1024, it requires ~130GB RAM.

**Comment by zhengruifeng** [comment]
@mengxr @WeichenXu123 
I am OK to revert this since it cause regression on high dimensional sparse datasets. Then we may also consider to do revert [LinearRegression](https://github.com/apache/spark/pull/27396) and [LinearSVC](https://github.com/apache/spark/pull/27360), since they three are impled in the same way.

**Comment by WeichenXu123** [comment]
@zhengruifeng Thanks! Also note your ongoing PR blockify GMM https://github.com/apache/spark/pull/27473 which do similar thing should also suspend for now.
We have found java-BLAS introduce significant JNI overhead and in some scenario introduce regression.

**Comment by zhengruifeng** [comment]
@WeichenXu123 Yes, I just mark GMM WIP. 
I am going to study this issue, thanks for pointing it out. I am sorry for failing to make a comprehensive test.

**Comment by zhengruifeng** [comment]
@WeichenXu123 @mengxr @srowen 
I just made a quick test on webspam:
I draw the first 10,000 sample from `webspam_wc_normalized_trigram.svm`, and the numFeatures=8,289,919 in the sampled dataset;
It's sparsity (percentage of non-zero values) is about 0.4489%.

This PR will fail dure to OOM in standardization, so I use a patch:
```scala
           val vec = features match {
              case dv: DenseVector =>
                var i = 0
                while (i < dv.size) {
                  val std = featuresStd(i)
                  if (std != 0) {
                    dv.values(i) /= std
                  } else {
                    dv.values(i) = 0.0
                  }
                  i += 1
                }
                dv
              case sv: SparseVector =>
                var j = 0
                while (j < sv.numActives) {
                  val i = sv.indices(j)
                  val std = featuresStd(i)
                  if (std != 0) {
                    sv.values(j) /= std
                  } else {
                    sv.values(j) = 0.0
                  }
                  j += 1
                }
                sv
            }

```

After that, I use following code to test performance:
spark-shell --driver-memory=32G --conf spark.driver.maxResultSize=4g
```scala
import org.apache.spark.ml.classification._
import org.apache.spark.storage.StorageLevel

val df = spark.read.format("libsvm").load("/data1/Datasets/webspam/webspam_wc_normalized_trigram.svm.10k").withColumn("label", (col("label")+1)/2)


val lr1 = new LogisticRegression().setMaxIter(100).setFamily("binomial").setBlockSize(128) // this PR
val start = System.currentTimeMillis; val model1 = lr1.fit(df); val end = System.currentTimeMillis; end - start


val lr2 = new LogisticRegression().setMaxIter(100).setFamily("binomial").setBlockSize(1024) // this PR
val start = System.currentTimeMillis; val model2 = lr2.fit(df); val end = System.currentTimeMillis; end - start


val lr = new LogisticRegression().setMaxIter(100).setFamily("binomial") // 2.4.4
val start = System.currentTimeMillis; val model = lr.fit(df); val end = System.currentTimeMillis; end - start

```

Result:

|Impl| This PR(blockSize=128) | This PR(blockSize=1024) | 2.4.4 |
|------|----------|------------|----------|
|summary.totalIterations | 31|31|31|
|duration | 298514 | 133982 | 108375 |
|RAM | 425 | 425 | 396 |


For this sparse dataset, this PR (with updated standardization) is about 23% slower, and use 7% more RAM.

So I aggre with you to revert this PR and relative PRs [LinearSVC](https://github.com/apache/spark/pull/27360), [LinearRegression](https://github.com/apache/spark/pull/27396).
Since [ALS/MLP extend HasBlockSize](https://github.com/apache/spark/pull/27389) depend on [LinearSVC](https://github.com/apache/spark/pull/27360), so may it also need to be reverted for now @huaxingao

**Comment by srowen** [comment]
Ahhh, OK. I didn't think enough about whether sparse vectors would behave significantly differently. Of course, should have been checked. I agree. I'm happy to merge a revert PR or @zhengruifeng you can too.

**Comment by huaxingao** [comment]
I can revert my PR https://github.com/apache/spark/pull/27389. But before I revert, I want to check with you folks.  My PR only has API changes:
1. add ```HasBlockSize``` to ALS so user can specify the blockSize for method ```blockify```
2. make MLP extend ```HasBlockSize``` in ```sharedParams.scala``` instead of having its own param ```blockSize```

It seems to me it's OK to keep these changes. Of course,  there is no ```HasBlockSize``` in ```sharedParams.scala``` any more after reverting https://github.com/apache/spark/pull/27360. I guess I can put ```HasBlockSize``` in ```sharedParams.scala``` instead of reverting my PR?

**Comment by mengxr** [comment]
I think we can keep the API only change and figure out a way to let the implementation automatically decides whether to blockify+densify for performance.

**Comment by zhengruifeng** [comment]
OK, I will revert those PRs, and then @huaxingao can add back **ALS/MLP extend HasBlockSize**  as a separate PR.
